shared/src/main/java/com/couchbase/lite/AbstractReplicator.java

```diff
@@ -361,13 +399,41 @@ public ListenerToken addChangeListener(Executor executor, ReplicatorChangeListen
     }
 
     /**
-     * Remove the given ReplicatorChangeListener from the this replicator.
+     * Remove the given ReplicatorChangeListener or DocumentReplicatedListener from the this replicator.
      */
     public void removeChangeListener(ListenerToken token) {
         synchronized (lock) {
-            if (token == null || !(token instanceof ReplicatorChangeListenerToken))
+            if (token == null || (!(token instanceof ReplicatorChangeListenerToken) && !(token instanceof DocumentReplicatedListenerToken)))
                 throw new IllegalArgumentException();
-            changeListenerTokens.remove(token);
+            if(docEndedListenerTokens.remove(token)==true) {
```

Comment:
I think this should check the size instead of the result of remove as below.

```Java
docEndedListenerTokens.remove(token);
if (docEndedListenerTokens.size() == 0) {
    setProgressLevel(ReplicatorProgressLevel.OVERALL);
}
```

Reply:
question, is setProgressLevel(ReplicatorProgressLevel.OVERALL); expensive? if (docEndedListenerTokens.size() == 0) check will be access many times even the doc Ended listener was never added...

Reply:
I’m not sure I understand your point. It’s not about performance here. The reason to set the progress level to overall when the size of docEndedListenerToken set is zero because there could be more than one listener added; set to overall should be done when all listeners are removed.